common/string_helpers_test.cpp

```diff
@@ -30,6 +30,7 @@ TEST(UnescapeStringLiteral, Valid) {
   EXPECT_THAT(UnescapeStringLiteral("\\x12"), Optional(Eq("\x12")));
   EXPECT_THAT(UnescapeStringLiteral("test", 1), Optional(Eq("test")));
   EXPECT_THAT(UnescapeStringLiteral("test\\#n", 1), Optional(Eq("test\n")));
+  EXPECT_THAT(UnescapeStringLiteral("r\\u{E9}al\\u{2764}\\u{FE0F}\\u{1F50A}!"), Optional(Eq("réal❤️🔊!")));
```

Comment:
Invalid test cases:
- `\u{e9}`
- `\u{E9z}`
- `\uz`
- `\u{`
- `\u`

The above test cases haven't been added to the program yet. Do let me know if I should add them.
I have added a few more requested test cases to the program now. They are valid.

Reply:
Yes, please add them. It's important that they're tested and handled correctly.

Reply:
Added these tests to another file called fail_unicode.carbon

Reply:
> ```
> // Various incomplete cases.
> "\\u"
> "\\uz"
> "\\u{"
> "\\u{z"
> "\\u{E9"
> "\\u{E9z"
> // Lowercase hex.
> "\\u{e9}"
> // Expecting this to be rejected because it's past the maximum codepoint.
> // ConvertCodePointToUTF8 should fail.
> "\\u{110000}"
> // More than 32 bits, but may incorrectly calculate to E9 when stored in a 32-bit unsigned.
> "\\u{FF000000E9}"
> ```
> 
> Also, one more valid case that I think can be tested:
> 
> ```
> // Maximum valid codepoint.
> "\\u{10FFFF}"
> // Nothing wrong with this, just a peculiar case.
> "\\u{0000}"
> ```

You seem to have missed several of these test cases, and importantly tests like `\\u{z` which I think the code still doesn't handle correctly. It's not ensuring there's a closing brace.

Test this _entire list_ of test cases. Testing edge cases is very important. Part of the reason I'm providing a list of test cases is to help you see and fix the bugs in your code; it's very important to add them when I'm asking you to.

Additionally, given #2040 and what you're testing, I'll note an additional test case that you need:

`"\\u000000E9"`

This should be valid, while the current code renders it invalid.

Reply:
Sorry about that, I must have missed those. I've added tests for all of them, except for two:

-For the zero codepoint `\u{0000}`, I'm having trouble writing a test case, since any attempt to put `\0` or any other form of "zero" in a C++ string just results in it terminating prematurely, because C++ uses 0 for string terminators. How should I write a test case for this one?

-For the `"\\u000000E9"`, I didn't add one exactly like that, since by my understanding of Carbon's Unicode semantics, there should be curly braces. I did add such a case with curly braces to the code, and I changed the logic in the unicode parsing code to check for a limit of 8 hex chars. Please let me know if those work ok.